File: 1750. Minimum Length of String After Deleting Similar Ends/minimumLength.js
/**
 * @param {string} s
 * @return {number}
 */
const minimumLength = (s) => {
  let left = 0;
  let right = s.length - 1;

  while (left < right && s[left] === s[right]) {
    while (left + 1 < right && s[left] == s[left + 1]) {
      left++;
    }

    while (left < right - 1 && s[right] == s[right - 1]) {
      right--;
    }

    left++;
    right--;
  }

  return right - left + 1;
};

const s = "cabaabac";
// const s = "ca";
// const s = "aabccabba";
console.log('minimumLength: ', minimumLength(s));

File: 1750. Minimum Length of String After Deleting Similar Ends/minimumLength.py
class Solution:
  def minimumLength(self, s: str) -> int:
    start = 0
    end = len(s) - 1

    while start < end and s[start] == s[end]:
      while start + 1 < end and s[start] == s[start + 1]:
        start += 1
      while start < end - 1 and s[end] == s[end - 1]:
        end -= 1

      start += 1
      end -= 1

    return end - start + 1
  
solution = Solution()
s = "cabaabac"
# s = "ca"
# s = "aabccabba"
print('minimumLength: ', solution.minimumLength(s))
